### PR TITLE
[test_nn] add custom device support for dropout tests、lazy_modules te…

### PR DESCRIPTION
**Pull request by summerdo**
add custom device support for dropout tests、lazy_modules tests and multihead_attention tests.

### CODE REVIEW

**Comment by summerdo** [comment]
@albanD sorry to bother, could you have a look?

**Comment by mikaylagawarecki** [comment]
@summerdo thanks for your patience, I will review this later today!

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -95,6 +95,10 @@
 from .composite_compliance import no_dispatch
 
 
+custom_device_mod = getattr(torch, torch._C._get_privateuse1_backend_name(), None)
+TEST_PRIVATEUSE1 = True if (hasattr(custom_device_mod, "is_available") and custom_device_mod.is_available()) else False
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_utils.py:99]: can you move this definition to where all the other `TEST_{BACKEND}`s are defined ([referring to this part](https://github.com/pytorch/pytorch/blob/210b5d0546ea6b9571c4b14952492fe4c466ab3d/torch/testing/_internal/common_utils.py#L1066-L1067))

File: test/nn/test_multihead_attention.py [hunk 2]
```diff
@@ -589,7 +589,8 @@ def test_multihead_self_attn_two_masks_fast_path_mock(self, device):
         Multihead self-attention should take fast path when both attention mask (mask type 0)
         and key padding mask (mask type 1) are provided at the same time on CPU and CUDA
         """
-        if device not in ['cpu', 'cuda']:
+        device = device.rstrip(':0123456789')
```
Comment (mikaylagawarecki) [inline test/nn/test_multihead_attention.py:592]: why this change?

File: test/nn/test_multihead_attention.py [hunk 3]
```diff
@@ -589,7 +589,8 @@ def test_multihead_self_attn_two_masks_fast_path_mock(self, device):
...
+        if device not in ['cpu', 'cuda', torch._C._get_privateuse1_backend_name()]:
```
Comment (mikaylagawarecki) [inline test/nn/test_multihead_attention.py:593]: nit: update skipTest below with 
`self.skipTest("Fastpath only runs on CPU and CUDA and PrivateUse1.")`

**Review by mikaylagawarecki — COMMENTED** [review]
This looks ok to me, other than some nits.

Separately, I want to check my understanding on the state of PrivateUse1 testing (is there an issue anywhere that summarizes this (?))  

(1) This PR updates `@onlyNativeDeviceTypes` to run tests on PrivateUse1
(2) This PR adds a decorator `@onlyCUDAAndPRIVATEUSE1` so that onlycuda tests can be run on privateuse1 as well 
(3) Separately `instantiate_device_type_tests` was already updated by https://github.com/pytorch/pytorch/pull/99960 to run for PrivateUse1

Is the above correct?

**Comment by summerdo** [comment]
> This looks ok to me, other than some nits.
> 
> Separately, I want to check my understanding on the state of PrivateUse1 testing (is there an issue anywhere that summarizes this (?))
> 
> (1) This PR updates `@onlyNativeDeviceTypes` to run tests on PrivateUse1 (2) This PR adds a decorator `@onlyCUDAAndPRIVATEUSE1` so that onlycuda tests can be run on privateuse1 as well (3) Separately `instantiate_device_type_tests` was already updated by #99960 to run for PrivateUse1
> 
> Is the above correct?
yes, your understanding is correct.

File: test/nn/test_multihead_attention.py (hunk 2 above)
Comment (summerdo) [reply test/nn/test_multihead_attention.py:592]: > why this change?

When running on a device such as cuda, the device will have a suffix card number(such as 'cuda:0'), thereby skipping the test case

File: test/nn/test_multihead_attention.py (hunk 3 above)
Comment (summerdo) [reply test/nn/test_multihead_attention.py:593]: > nit: update skipTest below with `self.skipTest("Fastpath only runs on CPU and CUDA and PrivateUse1.")`

updated

File: torch/testing/_internal/common_utils.py (hunk 1 above)
Comment (summerdo) [reply torch/testing/_internal/common_utils.py]: > can you move this definition to where all the other `TEST_{BACKEND}`s are defined ([referring to this part](https://github.com/pytorch/pytorch/blob/210b5d0546ea6b9571c4b14952492fe4c466ab3d/torch/testing/_internal/common_utils.py#L1066-L1067))

Modified as requested

**Comment by summerdo** [comment]
> This looks ok to me, other than some nits.
> 
> Separately, I want to check my understanding on the state of PrivateUse1 testing (is there an issue anywhere that summarizes this (?))
> 
> (1) This PR updates `@onlyNativeDeviceTypes` to run tests on PrivateUse1 (2) This PR adds a decorator `@onlyCUDAAndPRIVATEUSE1` so that onlycuda tests can be run on privateuse1 as well (3) Separately `instantiate_device_type_tests` was already updated by #99960 to run for PrivateUse1
> 
> Is the above correct?

File: test/nn/test_dropout.py [hunk 4]
```diff
@@ -50,11 +50,15 @@ def test_FeatureAlphaDropout(self):
         input = torch.randn(50, 20, 64, 64)
         self._test_alpha_dropout(nn.FeatureAlphaDropout, input)
 
-    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    @unittest.skipIf(not (TEST_CUDA or TEST_PRIVATEUSE1), "CUDA and PRIVATEUSE1 unavailable")
     def test_native_dropout_corner_case(self):
+        if TEST_CUDA:
+            devices = 'cuda'
```
[the remark this answers is not in the document]
Comment (summerdo) [reply test/nn/test_dropout.py]: Modified as suggested

**Review by mikaylagawarecki — APPROVED** [review]

**Comment by mikaylagawarecki** [comment]
Hm it looks like there is a merge conflict that needs to be resolved, could you rebase and resolve this please?

File: test/nn/test_multihead_attention.py (hunk 2 above)
Comment (mikaylagawarecki) [reply test/nn/test_multihead_attention.py:592]: ah, so the test was being erroneously skipped in CI before :o good catch
